### PR TITLE
[FSK-54]: Create Project Db Schema

### PR DESCRIPTION
**Pull request by pourdaavar**
- Model DB schema ✅
- Create the Prisma schema ✅
- Migrate DBS to the new schema  ✅
- Create a seed script ✅

NOTES: because of no permission, Database migration was not performed* - Done

### CODE REVIEW

File: prisma/schema.prisma [hunk 1]
```diff
@@ -38,3 +38,22 @@ enum FamilyStatus {
   DRAFT
   COMPLETE
 }
+
+model Project {
+  id String @id @default(cuid())
+
+  name        String
+  description String?
```
Comment (mhmda-83) [inline prisma/schema.prisma:46]: I think it's better to change String to Text because description can be long.

**Comment by mhmda-83** [comment]
I think you forgot to add sample project to seed script

**Review by mhmda-83 — CHANGES_REQUESTED** [review]

File: prisma/schema.prisma (hunk 1 above)
Comment (pourdaavar) [reply prisma/schema.prisma:46]: As Prisma [documentation](https://www.prisma.io/docs/concepts/database-connectors/postgresql#prisma-migrate) for text base datatype ( as string ), String was represented.
also this [docs](https://www.postgresql.org/docs/current/datatype-character.html)

**Comment by pourdaavar** [comment]
> I think you forgot to add sample project to seed script

No, it's not completed. I'm working on it.

File: .lintstagedrc.json [hunk 2]
```diff
@@ -3,5 +3,6 @@
     "prettier --write",
     "cspell --no-must-find-files"
   ],
-  "*.ts": ["eslint --fix", "cspell --no-must-find-files"]
+  "*.ts": ["eslint --fix", "cspell --no-must-find-files"],
+  "*.prisma": ["npx --no-install prisma format"]
```
Comment (MohammadForoutan) [inline .lintstagedrc.json:7]: I think It's good to have spell check  `.prisma` schema files too

File: prisma/seed.ts [hunk 3]
```diff
@@ -34,6 +34,13 @@ async function main() {
       `test family created, test family name is "${testFamily.name}"`,
     );
   }
+
+  const testProject = {
+    name: faker.lorem.word(5),
+    description: faker.lorem.sentence(3),
+  };
+  await prisma.project.create({ data: testProject });
```
Comment (MohammadForoutan) [inline prisma/seed.ts:42]: There should only be one project, so we can check if one already exists and decide not to create another

File: prisma/seed.ts [hunk 4]
```diff
@@ -34,6 +34,13 @@ async function main() {
...
+  console.log('New Project added');
 }
```
Comment (MohammadForoutan) [inline prisma/seed.ts:44]: we can print project name here

**Review by MohammadForoutan — CHANGES_REQUESTED** [review]
I commented on some points. check them please

File: prisma/seed.ts (hunk 3 above)
Comment (pourdaavar) [reply prisma/seed.ts:42]: I don't think so, as every time **seeding** occurs, we should **clean up the DB**, so there is no worry about **duplicate records**.

File: .lintstagedrc.json (hunk 2 above)
Comment (pourdaavar) [reply .lintstagedrc.json:7]: Good idea.

**Review by MohammadForoutan — APPROVED** [review]

**Review by ASafaeirad — APPROVED** [review]

File: prisma/schema.prisma (hunk 1 above)
Comment (mhmda-83) [reply prisma/schema.prisma:46]: yes that's correct, but you need to specify @db.VarChar for name because name can have limited size.

Comment (pourdaavar) [reply prisma/schema.prisma:46]: It's not better to do this validation on code instead of DB?

Comment (mhmda-83) [reply prisma/schema.prisma:46]: if we use VarChar it will take less storage on database

Comment (pourdaavar) [reply prisma/schema.prisma:46]: Please read the [docs](https://www.postgresql.org/docs/current/datatype-character.html) accurately.

> Don't specify a length maximum for database columns unless it is required by the application or you want to create a B-tree index on the column.
The performance will be the same, and PostgreSQL only allocates the space that is actually occupied by the data.

based on this [explaination](https://stackoverflow.com/questions/71396234/should-i-specify-a-size-to-a-postgresql-field-when-using-prisma)

**Comment by ASafaeirad** [comment]
There's a problem using Prisma and Cloud DBs.
For migrations, we need another DB instance.

File: prisma/schema.prisma (hunk 1 above)
Comment (mhmda-83) [reply prisma/schema.prisma:46]: but we want B-tree index on id (default index type for primary key is B-Tree)

**Review by ASafaeirad — APPROVED** [review]

File: prisma/migrations/20220918160012_dev/migration.sql [hunk 5]
```diff
@@ -0,0 +1,14 @@
+-- CreateEnum
+CREATE TYPE "ProjectStatus" AS ENUM ('PLANNING', 'IN_PROGRESS', 'SUSPENDED', 'DONE');
+
+-- CreateTable
+CREATE TABLE "Project" (
+    "id" TEXT NOT NULL,
```
Comment (mhmda-83) [inline prisma/migrations/20220918160012_dev/migration.sql:6]: id is primary key and primary key need to have b-tree index so it's required to set varchar data type for this.

Comment (pourdaavar) [reply prisma/migrations/20220918160012_dev/migration.sql:6]: PostgreSQL automatically creates an index for each unique constraint and primary key constraint to enforce uniqueness. Thus, it is not necessary to create an index explicitly for primary key columns.

Comment (mhmda-83) [reply prisma/migrations/20220918160012_dev/migration.sql:6]: but we need VarChar type for b-tree index. (I know it'll created automatically but as the docs that you sent said: you should not use text for a columns which uses B-Tree uindex)

**Review by MohammadForoutan — CHANGES_REQUESTED** [review]
*   the project name should be a string. we need `varchar` in index fields
*   migration is not updated

**Comment by pourdaavar** [comment]
> 

the project name was migrated on the previous [commit](https://github.com/fullstacksjs/the-charity-backend/pull/18/commits/9ec7de89f39f22ec2e4440d1b846e6be8073c277)

File: prisma/schema.prisma [hunk 6]
```diff
@@ -38,3 +38,22 @@ enum FamilyStatus {
   DRAFT
   COMPLETE
 }
+
+model Project {
+  id String @id @default(cuid()) @db.VarChar(32)
+
+  name        String  @db.VarChar(32)
```
Comment (MohammadForoutan) [inline prisma/schema.prisma]: name type should be a string (Text) as we decided

File: prisma/schema.prisma [hunk 7]
```diff
@@ -38,3 +38,22 @@ enum FamilyStatus {
   DRAFT
   COMPLETE
 }
+
+model Project {
```
Comment (MohammadForoutan) [inline prisma/schema.prisma:42]: model name should be lowecase

Comment (MohammadForoutan) [reply prisma/schema.prisma:42]: add `@@map("project")` at end of the model to change its name in DB

**Review by MohammadForoutan — APPROVED** [review]

**Pull request merged**